### PR TITLE
Admin Comments section

### PR DESCRIPTION
**Pull request by innatewonderer**
- [x] Setup pagy gem for pagination https://github.com/sudara/alonetone/issues/247
- [x] setup basic admin namespace https://github.com/sudara/alonetone/issues/266 
- [x] setup admin/comments routes/controller/view and specs https://github.com/sudara/alonetone/issues/267
- [ ] setup admin/user routes/controller/view and specs https://github.com/sudara/alonetone/issues/268
- [ ] setup admin/asset routes/controller/view and specs https://github.com/sudara/alonetone/issues/269

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -94,6 +94,10 @@ def is_deliverable?
   def truncate_user_agent
     self.user_agent = user_agent.try(:slice, 0, 255)
   end
+
+  def body_snippet
+    self.body.truncate(100)
```
[the remark this answers is not in the document]
Comment (sudara) [reply app/models/comment.rb]: `self` is only needed on assignment — but also, we probably want to use Rails' view helper https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-truncate

File: app/controllers/admin/comments/mark_as_spam_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class Admin::Comments::MarkAsSpamController < Admin::BaseController
```
[the remark this answers is not in the document]
Comment (sudara) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: This bit us on another project recently, basically Rails' can't code load properly without explicit nesting because it might not have any idea what `Comments` is, a module or class. More here https://edgeguides.rubyonrails.org/autoloading_and_reloading_constants.html#nesting

[the remark this answers is not in the document]
Comment (sudara) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Oh, sorry, just saw this was `Outdated` so I'll dismiss.

File: Gemfile [hunk 3]
```diff
@@ -69,6 +69,10 @@ gem 'skylight'
 gem 'sidekiq'
 gem 'dalli'
 
+# admin functionality
+gem "datagrid"
```
Comment (sudara) [inline Gemfile]: Can we remove `datagrid` until we identify an explicit need on the frontend? I prefer to be very pessimistic on dependencies when possible.

File: app/controllers/admin/comments/add_to_blocklist_controller.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module Admin
+  module Comments
+    class AddToBlocklistController < Admin::BaseController
```
Comment (sudara) [inline app/controllers/admin/comments/add_to_blocklist_controller.rb]: Couple thoughts, you can hit me up in slack to hash it out:

1. I think we'll need to add/remove/index the blocklist (so the controller should be probably named `BlocklistController`)

2. It's IPs that we are managing (not comments) and we'll also be blocking them based on tracks, users, etc — so I'm thinking we probably want it to be `Admin::BlocklistController`

File: app/controllers/admin/comments/mark_as_spam_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+module Admin
+  module Comments
+    class MarkAsSpamController < Admin::BaseController
```
Comment (sudara) [inline app/controllers/admin/comments/mark_as_spam_controller.rb]: What do you think about having this just be in `Admin::CommentsController`? 

I guess the question is if we are doing something RESTy or not. If so, we'd want to name this `Admin::Comments::SpamController` (since we'll be doing more than marking as spam, we'll be unspamming things too). But I'm leaning towards thinking we won't have anything REST-like here, I can only think of these actions right now:

1. Listing comments
2. Marking one as spam
3. Marking one as ham
4. Marking all from an ip as spam

File: app/controllers/admin/comments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module Admin
+  class CommentsController < Admin::BaseController
+    def index
+      @pagy, @comments = pagy(Comment.all)
```
Comment (sudara) [inline app/controllers/admin/comments_controller.rb]: Woo pagy!

File: app/models/comment.rb [hunk 7]
```diff
@@ -90,6 +90,10 @@ def is_deliverable?
   def truncate_user_agent
     self.user_agent = user_agent.try(:slice, 0, 255)
   end
+
+  def body_snippet
+    body.truncate(100)
```
Comment (sudara) [inline app/models/comment.rb]: It's probably a bit nicer to use Rails' truncate stuff as a view helper: https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-truncate

File: app/controllers/admin/comments/mark_as_spam_controller.rb (hunk 5 above)
Comment (sudara) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Hmm, and then I guess I'm dreaming of a button that says "Delete All Spam older than 30 days"

File: Gemfile (hunk 3 above)
Comment (innatewonderer) [reply Gemfile]: Oh yep, Leftover from when I was experimenting with using it.

File: app/controllers/admin/comments/mark_as_spam_controller.rb (hunk 5 above)
Comment (innatewonderer) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Yes, I've been trying to be more RESTful. I think this way will force us to keep files small and specific... but OK to go either way honestly.

File: app/controllers/admin/comments_controller.rb (hunk 6 above)
Comment (innatewonderer) [reply app/controllers/admin/comments_controller.rb]: I'm going to do some research on how to actually test performance for this :) Shouldn't be too complicated to switch other controllers/views to this just time and 👀 .

File: app/controllers/admin/comments/mark_as_spam_controller.rb (hunk 5 above)
Comment (innatewonderer) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Also, @sudara ... `Marking one as ham` ... you telling me `ham` is not a typo? :)

Comment (sudara) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Thinking about the REST thing.... In order for REST to make sense, we'd need identify the entity/element we are operating on in collection/singular. `MarkAsSpam` doesn't work for me as entity we are listing/creating/destroying; it's not a noun that we are going to perform actions on.

`Comment` could be the element (`Admin::CommentsController`) and we could use member/collection routes to add actions that adjust the comment's attributes https://guides.rubyonrails.org/routing.html#adding-more-restful-actions — this is my preferred route because "spam" isn't an entity, it's an attribute that we are modifying, so it makes intuitive sense.

It could also be `Admin:Comments::Spam` if you want to view `Spam` as the noun being created/destroyed. I'd argue this would end up less clear to understand and read. What would marking something as "not spam" be? Using `DELETE` for that would be like  `destroying a comment spam` which is hard to understand I think.

But maybe there's a noun that would work? In this blog post, they talk about going more granular than the data model behind the API, one of the examples is `ChangeOfAddress`: https://www.thoughtworks.com/insights/blog/rest-api-design-resource-modeling

Hmmm, something like `MarkedSpam` might work? (a bit clunky). 

The action being performed as 1) informing an external API that they mis-categorized a comment (as "spam" or "not spam") and 2) changing the visibility of the comment on the site. I dunno, to me these keep seeming like `PUT` actions on a `Comment` resource, but maybe you can find a noun that works?

Bottom line is I'm always fine with going more granular given that we "list/create/update/destroy" a entity/resource/noun (otherwise we aren't actually doing REST)

> you telling me ham is not a typo? :)

Hehe, yeah, I actually don't like "ham" as the opposite of spam, it's confusing, but I guess it's a thing? 🐷 `mark_as_spam` and `mark_as_not_spam` sound better 🍡

File: app/controllers/admin/comments_controller.rb (hunk 6 above)
Comment (sudara) [reply app/controllers/admin/comments_controller.rb]: OMG, I'm too excited. I'd be curious just to see what kinds of requests it makes (as seen in logs)

File: app/models/comment.rb (hunk 7 above)
Comment (sudara) [reply app/models/comment.rb]: My gut says since the truncate helper is there, let's use it. I don't intuitively understand why we'd want to add code/abstraction to accomplish the same end result, but take that with a grain of salt and feel free to explain! I'm sorta a vanilla-rails person who prioritizes claritty and straightforwardness and gets confused easily with patterns and so on. My day job is basically unravelling patterns people add to their rails apps to regain performance/clarity :D I do think there are a few of places in alonetone in need of encapsulation (for example user names / profile pictures)...

File: app/controllers/admin/comments/mark_as_spam_controller.rb (hunk 5 above)
Comment (sudara) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Oooh, maybe `SpamReport` or `SpamStatus`? Still might be not really REST-ful because we'd actually be listing comments, not SpamReports, and we'd never update or destroy a SpamReport.... hm...

File: app/controllers/admin/comments_controller.rb (hunk 6 above)
Comment (innatewonderer) [reply app/controllers/admin/comments_controller.rb]: Well, for starters :) we'd go from this:
```
  <div id="left" class="comments_area box">
    <h1>Recent Comments</h1>
    <div class="mini_paginator">
      <%= will_paginate @comments, param_name: :page%>
    </div>
    <%= render partial: 'shared/comment', collection: @comments %>
    <div class="mini_paginator">
      <%= will_paginate @comments, param_name: :page%>
    </div>
  </div>
```
to this:
```
  <div id="left" class="comments_area box">
    <h1>Recent Comments With Pagy</h1>
    <%== pagy_nav @pagy %>
    <%= render partial: 'shared/comment', collection: @comments_for_pagy %>
    <%== pagy_nav @pagy %>
  </div>
```

Comment (innatewonderer) [reply app/controllers/admin/comments_controller.rb]: Don't quote me on that, but here is what I did:
- setup some caching in my `development.rb` (it seems like pagy's benefits increase quite a lot once caching is enabled)
- performed simple Rails view benchmarking test 
Hard reloads: 
```
[22:35:16] (pagy-test) alonetone
$ tail -f log/development.log | grep 'Comments'
Processing by CommentsController#index as HTML
Comments (4156.4ms)
Comments with pagy (141.5ms)
Processing by CommentsController#index as HTML
Comments (1271.1ms)
Comments with pagy (111.4ms)
Processing by CommentsController#index as HTML
Comments (3525.9ms)
Comments with pagy (1504.4ms)
Processing by CommentsController#index as HTML
Comments (1607.2ms)
Comments with pagy (89.4ms)
Processing by CommentsController#index as HTML
Comments (376.9ms)
Comments with pagy (115.0ms)
```
Soft reloads
```
[22:36:52] (pagy-test) alonetone
$ tail -f log/development.log | grep 'Comments'
Processing by CommentsController#index as HTML
Comments (2595.4ms)
Comments with pagy (91.5ms)
Processing by CommentsController#index as HTML
Comments (1567.8ms)
Comments with pagy (188.0ms)
Processing by CommentsController#index as HTML
Comments (3659.9ms)
Comments with pagy (1039.4ms)
Processing by CommentsController#index as HTML
Comments (2545.5ms)
Comments with pagy (199.1ms)
Processing by CommentsController#index as HTML
Comments (643.9ms)
Comments with pagy (140.0ms)
Processing by CommentsController#index as HTML
Comments (2331.8ms)
Comments with pagy (103.9ms)
```

File: app/controllers/admin/comments/mark_as_spam_controller.rb (hunk 5 above)
Comment (innatewonderer) [reply app/controllers/admin/comments/mark_as_spam_controller.rb]: Sorry didn't push the code earlier :) I am already convinced on member routes and `def spam/ham`
